xapi-project/sm
# CA-171849: Symlink in /dev/disk/by-scsid failed for dm devices

XS ships with its own udev rules for multipath and scsi devices.
It is possible that udev's behaviour changed but the net result now
is that we are unable to create a symlink for dm devices in /dev/disk/by-scsid.

/dev/disk/by-scsid is a custom dir we for standard scsi devices and it is
correctly populated but when it comes to dm devices built on top of those
(as it is the case for multipath devices), the link is not there.

The reason being, dm creation is always a 2 steps process
(from udev point of view):
the device is added (action = add) empty and then its status changes
(action = change) to make it point to the scsi devices to use for multipathing.

The rule contained in sm creates the link only during add.
In fact, the symlink is there after addition (you can test this using
"udevadm test") but when the change action kicks in, the link is removed
due to some internal udev rules about symlink management.

The fix is just to create the symlink regardless of the action
(when the device is removed udev is clever enough to remove the link itself).

Easy way to test it:

* create multiple connections to the same device (e.g. different paths to same iscsi target)
* check those device are listed under /dev/disk/by-scsid/<scsid>/ (they are)
* create a multipath device on top of them
* check a corresponding m-* device is created
* check there is now a "mapper" entry in the above dir

The last step fails without the fix.

Signed-off-by: Germano Percossi <germano.percossi@citrix.com>
